common/raw_hashtable.h

```diff
@@ -1244,7 +1203,79 @@ template <typename LookupKeyT>
   // And lastly insert the lookup_key into an index in the newly grown map and
   // return that index for use.
   --growth_budget_;
-  return InsertIntoEmpty(lookup_key);
+  return InsertIntoEmpty(lookup_key, key_context);
+}
+
+template <typename InputBaseT, ssize_t SmallSize>
+TableImpl<InputBaseT, SmallSize>::TableImpl(const TableImpl& arg)
+    : BaseT(arg.alloc_size(), arg.growth_budget_, SmallSize) {
+  CARBON_DCHECK(arg.small_alloc_size_ == SmallSize);
+
+  ssize_t local_size = arg.alloc_size();
+
+  if (SmallSize > 0 && arg.is_small()) {
+    CARBON_DCHECK(local_size == SmallSize);
+    this->storage() = small_storage();
+  } else {
+    this->storage() = BaseT::Allocate(local_size);
+  }
+
+  uint8_t* local_metadata = this->metadata();
+  EntryT* local_entries = this->entries();
+  const uint8_t* local_arg_metadata = arg.metadata();
+  const EntryT* local_arg_entries = arg.entries();
+  memcpy(local_metadata, local_arg_metadata, local_size);
```

Comment:
Say that we are preserving which slot each entry is in, along with all the tombstones, instead of rehashing just the present values.